### ISSUE TITLE
 Tooltips gets in the way of general operations in the editor

### ISSUE THREAD

**Issue by pflannery** [microsoft/vscode#15667]
- VSCode Version: 1.7.1
- OS Version: Windows 10 64bit

Steps to Reproduce:

![hover-pains](https://cloud.githubusercontent.com/assets/1727302/20403028/15085e66-acf7-11e6-94cc-1ca1426ad375.gif)

I'm finding that the hover over tooltip gets in the way of general operations in the editor. i.e. when trying to double click on words. The tooltip prevents selection (see the screenshot).

Can we have a setting to change the delay timing for these tooltips?

**Comment by alexandrudima** [comment]
It is not currently possible to configure the delay until the hover is shown. It is however possible to disable it via `"editor.hover": false`

**Comment by mfrost66** [comment]
+1 some poslish on usablity would be wonderful

**Comment by DSoa** [comment]
I don't want to see popups on everything I hover over, but do want them when there is an error (lint, etc.). Setting editor.hover to false turns off all hover help.
Can you make it more granular?  Or can you add a control-key+hover/click or something to show the tooltip?

**Comment by chrisdothtml** [comment]
> It is however possible to disable it via "editor.hover": false

@alexandrudima Was this setting removed? It apparently doesn't exist. The hints hinder more than help me and I would really like to disable all of them. Disabling `parameterHints` didn't solve it either

**Comment by freMea** [comment]
`"editor.hover": false` seems to work although **Code** returned the settings as
`Unknown configuration setting`.

Note it won't work as per language setting, it has to be a general setting.

This need to be fixed and have more granularity on setting this: delay, what tooltip to disable, per language setting.

In the meantime, you still can use `Ctrl+k, Ctrl+i` to show the tooltip at the caret position.

_setup: v1.10.2 on Win x64 pro_

**Comment by chrisdothtml** [comment]
@freMea That appears to have done the trick for Mac v1.10.2 as well. Weird

**Comment by freMea** [comment]
Beware that if `"editor.hover": false` is used and you intend to use `Ctrl+k, Ctrl+i` to force tooltip display at caret position, you may encounter an error that will prevent tooltip to display in any case after **VS Code** has restarted like
`Cannot read property 'startShowingAt' of undefined`
To fix this, you have no choice but set `"editor.hover": true,` and restart **Code**.

Thus `"editor.hover": false` is far from being ideal since it will break the tooltips even when you try to force them to show up.

**Comment by mikeatgithub** [comment]
The best solution to avoid the "hover over tool-tip" annoyance would be to assign its invocation to a keyboard combination like "CTRL+SHIFT+T". This way it is always off until you need it at which time you could invoke it by hitting "CTRL+SHIFT+T" or a keyboard combination of user's choice settable in "settings.json" file.

**Comment by bpasero** [comment]
@alexandrudima any reason why `editor.hover` is not a public setting? I have a user that searched for it and wants to enable.

**Comment by alexandrudima** [comment]
@bpasero You're going to need to negotiate with @isidorn The `editor.hover` is programmatically disabled when debugging and programmatically enabled when debugging. Exposing the setting publicly would mean for the debugger code paths to read the setting from the settings service according to our language-based options to re-establish the user setting [here](https://github.com/Microsoft/vscode/blob/2c85446dc85f8eaf96e101a78abd062a4094d7cd/src/vs/workbench/parts/debug/electron-browser/debugEditorContribution.ts#L278).

**Comment by isidorn** [comment]
@alexandrudima you describe it quite well, nothing to add here except please create a feature request for this and assign it to me. If there are users wanting this then we can surely add it.

**Comment by bpasero** [comment]
Ok I opened https://github.com/Microsoft/vscode/issues/28535

**Comment by madprops** [comment]
I agree with @DSoa 

>Can you make it more granular? Or can you add **a control-key+hover**/click or something to show the tooltip?

**Comment by wmadden** [comment]
A scenario occurring frequently for me is that the hover comes up while typing, and there's no way to hide it without using the mouse to click somewhere else. It's particularly frustrating when it's useless generated type information - which is enormous - like this:

![omghover](https://user-images.githubusercontent.com/12165/33549927-72548c12-d8ec-11e7-8304-4278e0da78ce.gif)

I would *love* a simple key binding to close the hover information. There's a keybinding for `editor.action.showHover` but there's no `editor.action.hideHover`!

**Comment by pakastin** [comment]
![nayttokuva 2017-12-21 kello 12 33 45](https://user-images.githubusercontent.com/1475902/34253676-4f61cb1c-e651-11e7-9382-e1c2c3d675da.png)

Yup, super annoying!

**Comment by itsjtwright** [comment]
Yeah I am having the same issue as @pakastin and @wmadden , a way to choose a cap to character count on the tool tip before going to "..." would be great

**Comment by estorski** [comment]
One thing that would solve this "tooltip issue" for me is to make it NOT hit-testable so that the tooltip does not get in my way every now and then when I click the code.

**Comment by usernamehw** [comment]
@wmadden @pakastin This is not hover. It's `parameter info` or `parameter hints`
```json
"editor.parameterHints": false,
```

**Comment by dnutels** [comment]
None of these work in Insiders. Literally do not do anything.

**Comment by pdecat** [comment]
Current workaround with 1.20.0 debian package:

```
# sudo sed -i s/HOVER_TIME=300/HOVER_TIME=1000/ /usr/share/code/resources/app/out/vs/workbench/workbench.main.js
```

**Comment by rob3c** [comment]
Still going since 2016 - impressive! :-)

I'm just wondering how this isn't enough of a daily annoyance for the team members who actually know the implementation that it isn't resolved already, by necessity of maintaining sanity. Are they all using the vim extension exclusively, so they never have a need to click to focus in some code and be routinely prevented by the tooltip? It's always been a problem, but it seems particularly bad recently, which is why I came digging through the issue tracker. Is it really so complicated, that exposing a custom delay or required shortcut key hook is um, not an option? I like them sometimes, so disabling them isn't desired - just controlling them better somehow.

**Comment by theRemix** [comment]
hitting `shift` + `esc` hides the tooltip.

**Comment by jaxelsson** [comment]
I think more granularity on usability would be very welcomed.  

1. Something as simple as a longer delay (or delay setting) would go a long way to keeping the hyperactive feeling black boxes from constantly obscuring your direct actions and being helpful again.  
2. In addition, lack of context granularity really becomes annoying when you're trying to select text.  If you have text selected, it's a lot less likely that you still want the tooltips to pop up at all.  
3. Maybe it is possible to adjust the placement or styling of the black box to be a little less of a "close talker" intruding on my attention, so it's not so uncomfortable and distracting.

For me this is distracting and obnoxious enough that I start considering switching to something else entirely.  Which is sad, so I persist... for now.

**Comment by jaxelsson** [comment]
I think all in all more granularity on usability would be very welcomed.  

1. Something as simple as a longer delay (or delay setting) would go a long way to keeping the hyperactive feeling boxes from being obscuring and annoying to being helpful again.  
2. Lack of context granularity really becomes annoying when you're trying to select text as well.  If you have text selected, it's a lot less likely that you still want the

**Comment by defields923** [comment]
I've had this issue since I first started using vscode, and wow am I surprised to find there isn't a solution??

The biggest annoyance is the issue wmadden brought up. typing multiple key/value pairs in an object while the tooltip pops up every new line, preventing me from using my arrow keys to jump lines, forcing me push esc to get rid of it. >(

I would like an option to selectively hide certain tooltips (because I do rather like the suggestions tooltip), or maybe the ability to move it to another part of the page, maybe upper-right corner.

**Comment by proimage** [comment]
These 4 changes would fix this entirely.

1.  Only keep the tooltip open when the mouse is over the triggering code string itself. Hide using the usual delay once the mouse moves over the tooltip window.
2. Clicks on the tooltip window should pass-through to the code beneath.
3. Holding a modifier key (Shift? Ctrl?) to negate both 1 and 2 in case someone wants to select a code sample from the tooltip window itself.
4. Differentiate between informational tooltips (the ones complained about here) and code-completion prompts, and have separate "showAfter" settings for each.

**Comment by tushar-singh** [comment]
There should definitely be a way to at least disable the arrow navigation inside tooltips. I'd much rather be able to navigate through my code than endless tooltips

**Comment by kellytk** [comment]
Not being able to configure the hover delay introduces substantial friction to my use of VS Code.  Additionally a key binding to hide the view, such as esc, would be helpful.

**Comment by alexandrudima** [comment]
I plan to look into improving things for the **hover**.

Just to clarify on expectations, this issue has gotten bombarded with various ideas or complaints about two different widgets. One is the hover and that shows typically when using the mouse over a symbol and the other is the parameter hints widget and that shows when typing (similar to suggestions) and renders the arguments that should be passed in to a method call. I plan to improve things for the **hover**.

| Type | Screenshot | Link |
|---|---|---|
| **Hover** | ![hover-pains](https://cloud.githubusercontent.com/assets/1727302/20403028/15085e66-acf7-11e6-94cc-1ca1426ad375.gif) | https://github.com/Microsoft/vscode/issues/15667#issue-190133079 |
| **Parameter Hints** | ![omghover](https://user-images.githubusercontent.com/12165/33549927-72548c12-d8ec-11e7-8304-4278e0da78ce.gif) | https://github.com/Microsoft/vscode/issues/15667#issuecomment-348932383 |
| **Parameter Hints** | ![nayttokuva 2017-12-21 kello 12 33 45](https://user-images.githubusercontent.com/1475902/34253676-4f61cb1c-e651-11e7-9382-e1c2c3d675da.png) | https://github.com/Microsoft/vscode/issues/15667#issuecomment-353326281 |

---

To summarise the current proposed ideas regarding **hover**:

* The setting `"editor.hover"` should no longer be an internal thing and it should work per language.
* Add option to configure the duration to hover until something is painted.
* Add option to show only errors & warnings in hover and not compute semantic hovers.
* Add option to configure hover to be shown with a different gesture than mouse hovering. (e.g. <kbd>Ctrl</kbd> key pressed and hover).
* If `"editor.hover": false`, the hover cannot be shown via keybindings.
* Allow clicking to pass through the hover (i.e. hover should not eat mouse events).
* Do not show hover when selecting text
* Hide the hover when the mouse leaves the triggering code.

**Comment by eamodio** [comment]
All sounds great, except for
> Allow clicking to pass through the hover (i.e. hover should not eat mouse events).

While I think this is probably a good default, I often copy from hovers, so I would request a way to allow the hover to have focus. Maybe Ctrl+click or something. So we don't loose that ability.

**Comment by michael-38** [comment]
Until someone has a better suggestion, I'm just going to set the delay to a higher number...

```"editor.hover.delay": 3000000```

**Comment by proimage** [comment]
BTW, I think disabling the hover also affects the color selector popup in CSS files...

**Comment by reeddunkle** [comment]
I believe that `"editor.hover": false` no longer works.

**Comment by reeddunkle** [comment]
@chrisdothtml You're right. I guess the old setting `"editor.hover": false` was preventing the others from working. The delay property works as expected now too. Thanks!

**Comment by Nodws** [comment]
Ah this bad boy has saved me from frustration `"editor.hover.delay": 3000`

**Comment by mblandfo** [comment]
I didn't see it on @alexandrudima 's list, but right now the hover appears **directly above** the hover element. I'd suggest adding 20px or 30px margin so it's a little less likely to get in the way.

"editor.hover.delay" seems to default to 300, I think 500 may be good enough. With 300, as I mouse up the screen, and I slow down when I get close to the text I want to select, as the mouse passes over the line below the hover for that line shows up... right on top of the text I was about to select!

**Comment by jasel-lewis** [comment]
Any movement on the "Parameter Hints" issue @wmadden presented? ([comment link](https://github.com/Microsoft/vscode/issues/15667#issuecomment-348932383))

As his GIF demonstrates, this is _not_ a hover-related issue, rather, the tooltip appears as you type.  ESC works to clear it, but the very next keystroke, the tooltip re-appears.  And it's VERY inconsistent, at least for me, in when it displays.  I've yet to determine the rhyme or reasoning.  At times, it even appears when typing a comment line...

**Comment by edwarddim** [comment]
"editor.hover.enabled": false

It worked for me

**Comment by jasel-lewis** [comment]
@edwarddim I'm glad `"editor.hover.enabled": false` works for whatever concern you've experienced.  

The issue which @wmadden has [described in his post](https://github.com/Microsoft/vscode/issues/15667#issuecomment-348932383), and to which I am referring, is _not_ related to a _mouse_ hover condition.  The graphic @wmadden provides in his post is exactly what I am experiencing.  The pop-up _appears_ the same as a hover action, but it occurs without ever moving the mouse - it occurs when _typing_ and/or using nothing more than the keyboard's arrow keys to move around in the code.

I'll restate that `"editor.hover.enabled": false` _does not_ prevent that which is depicted in @wmadden's post.

**Comment by red-lonely-neuron** [comment]
I have Visual Studio Code 1.26.1 with Electron 2.0.5 on Linux and I have attempted to change the value of **editor.hover.delay** to **2000** (something different than the default 300) in the user settings. The result was the tooltips appeared after 2000 ms, but after restarting Code, the delay was again 300 (I guess since the tooltips appeared too fast). I have also attempted to set the value in the workspace settings and I've got the same result: after a restart, Code doesn't seem to take into account the value of **editor.hover.delay** in the user settings or in the workspace settings. It takes it into account only after modifying it to something else, then setting it to the desired value.

**Comment by reeddunkle** [comment]
@red-lonely-neuron I previously had the setting `"editor.hover": false` (before it was deprecated), and found that that setting was preventing the editor from registering the new hover delay setting `"editor.hover.delay": 3000` I was attempting to add.

So there's a chance some other setting is conflicting with your hover delay setting.

**Comment by red-lonely-neuron** [comment]
@reeddunkle: there's no other modified **editor.hover** setting either in the user settings file, or in the workspace one. There is a **editor.formatOnSave** set to false.

**Comment by reeddunkle** [comment]
@red-lonely-neuron I don't think that `"editor.formatOnSave"` would interfere. Sorry, I don't have any other ideas.

**Comment by alexandrudima** [comment]
@joaomoreno did we remove the setting `editor.parameterHints` for toggling on/off parameter hints ?

**Comment by joaomoreno** [comment]
@alexandrudima It seems to have changed to `editor.parameterHints.enabled`: https://github.com/Microsoft/vscode/pull/55354 cc @jrieken

**Comment by jakepark** [comment]
Thank you both. I can confirm that this fixes the parameterHints display issue.

`"editor.parameterHints.enabled": false`

Tooltips and its discussion is still relevant for this ticket.

**Comment by benzvan** [comment]
Any progress on this? I'm just starting to try using vscode and I'm finding it very difficult with the parameterHints popping up and covering my code after I've already dismissed the parameter hint for the block of code I'm working on.

**Comment by pakastin** [comment]
I just disabled every single tooltip I could find 🤷‍♂️

**Comment by benzvan** [comment]
I actually find the tooltips super helpful when they aren't blocking the actual code I'm writing.

**Comment by sylv3r** [comment]
this parameterHints issue is bothering me for a while now, up to the point I ended up here to find a solution. 
Being able to turn it off completely is not really solving the issue imho, that's hiding the problem (but helps for sure with the trade off to loose useful hints).

One solution I can think of, is to let one set where he wants this popup to appear, something like `"editor.parameterHints.position"` with options like `cursor`, `topLeft`, `topRight`, `bottomLeft`, `bottomRight`, etc 

That way we could still enjoy the hints without having them overlaying the code we're working on.

**Comment by jakepark** [comment]
In the past I've coded modals that do exactly as you describe, @sylv3r . 

As part of your feature request, please provide wireframes, some user interaction flow (gif's are worth a million points), edge case solutions (e.g. what should happen when the cursor is at the extremities of the screen), and other potential problems I haven't thought of off the top of my head :)

Not saying I'll work on it but hopefully this helps describe the problem complexity. 

Honestly since I turned off the feature a year ago, I feel like I've opened more library code to fully understand dependencies being injected..

**Comment by proimage** [comment]
Might I suggest implementing modifier keys to control the behavior of these popups and hint windows and whatever the heck you call them? For example, by default, have these popups visible but non-interactive (not clickable). You have to hold `CTRL`/`CMD` to be able to interact with the popups—that's one modifier key. Another one would be `SHIFT` perhaps, which would make the popup transparent while the key is pressed.

This general "Hold a modifier key to interact with something" pattern is already established for when you click on links in those popups, so why not extend it to affect the interactivity behavior of the entire popup?

**Comment by affordablequalityhomes** [comment]
I would like them to be unobtrusive by default, without having to set some custom setting which may have unintended side-effects. Can they be a little transparent, and not clickable? But then if you do `CTRL + .` they are? Just brain storming. Something like that. So the people who want to do something with the tooltip itself still have a way, but 99% of the time when you don't, they don't block you clicking or selecting text or using the editor. Thanks.

**Comment by freMea** [comment]
A bar/panel at the bottom of the editor that would appear with no delay could do the trick too.

But with the new flexible layout released in may 2020, a parameter info view that one can move everywhere would be nice too. Maybe a new activity being oriented towards code documentation that would implements the current parameter info tooltip as well as other documentation aggregator like DevDocs would be nice and complete.

I have a workaround to have access to this info tooltip without being annoyed by them. I explain it here #102198 because this highlighted another bug.

Here are other related issues more or less duplicate : #33752, #16221 (an awesome feature suggestion), #51253, #30797, #63144.

**Comment by chad-autry** [comment]
I endorse the suggestion to make the default "editor.hover.delay" greater than 300 ms.
It ironically prevents double cliking un-used variables to copy the name and paste elsewhere, because it pops up a tooltip about them being unused.

**Comment by aminomancer** [comment]
can we get a setting to change the anchor position for the hover tooltips? I actually like the 300ms delay, but I'm forced to change it to 700 because it gets in my way. If you could just move it so it doesn't cover up the thing you're clicking on, then the short delay wouldn't matter. Especially with web languages because there's an MDN documentation link at the bottom of the tooltip. So say I'm trying to click on a line but I end up hovering over a CSS property that's beneath it. Right before I click, the tooltip pops up, and now my cursor is right over the MDN documentation link. So now instead of clicking in the editor I'm opening a firefox window.

This happened to me on a daily basis until I just bit the bullet and increased the delay. But now I find the delay irritating too lol. A simple solution is to just make the tooltips anchored to the *top right* of the hovered context. And not just moving it to the right but move it up a little bit too. Right now it actually seems to overlap with the hovered line by like 3 or 4px. Instead, the lower edge of the tooltip should run halfway through the line above the hovered line. That way there's still room to click on the line above the hovered line.

Ideally this would be a preference so users can choose between anchor points as well as a fixed position. Some users might prefer to have the tooltips completely out of the way of the workspace. Well, a fixed tooltip would still be in the workspace, but it wouldn't be in the immediate vicinity of the cursor, which means it wouldn't be in the immediate vicinity of the code the user is currently working on. I might actually prefer that myself, but since the option doesn't exist, I haven't tried it yet. Right now I think my preference would just be to move the tooltip up and to the right a little bit. Or to anchor it to the bottom of the hovered element instead, and then move it down and to the right. That's how tooltips are typically implemented in browsers.

**Comment by justsml** [comment]
I don't think this problem is solved by changing the position OR the duration. 
The issue is the tooltip covers content when trying to interact or read it.

[This person has a fantastic suggestion:](https://github.com/Microsoft/vscode/issues/15667#issuecomment-562273298)
> hold `CTRL`/`CMD` to be able to interact with the popups—that's one modifier key. Another one would be `SHIFT` perhaps, which would make the popup transparent while the key is pressed.

Another idea (with possibly less conflicts) is to support a "double tap" for modifier keys. 

Double tap shift to toggle the tips on/off would be amazing for my workflow.

**Comment by justsml** [comment]
Could the click behavior on tooltips be 'fixed' with an option to apply some CSS like  `pointer-events: none;`? (Then that can be toggled with a modifier key or double-tap modifier.)

Just brainstorming... 😇

**Comment by jsmaniac** [comment]
> Could the click behavior on tooltips be 'fixed' with an option to apply some CSS like `pointer-events: none;`? (Then that can be toggled with a modifier key or double-tap modifier.)

I thought about that, but it would be impossible to select text inside the hover (which can contain useful stuff like the types of what's under the cursor) :-/ . Also I'm not sure scrolling would still work with `pointer-events: none`

**Comment by justsml** [comment]
I think I used the wrong word "option."

I'm not sure what options might be required, but I've got 2 rough behaviors in mind:

1. The tooltips show as configured, however in order to interact (click or scroll) I'd hold down a modifier key `[ctrl]` (or similar.) Any click on the tooltip would effectively click the item immediately beneath as though the tooltip wasn't there. Holding the modifier would be the only way to scroll, or click on links.

2. Another approach that might work: add ability to rapidly toggle the tooltips on/off. Where the tooltips wouldn't show on-hover at all, until I double tap `[shift or ctrl]`. Then they would stay enabled until I toggle them away.

**Comment by aminomancer** [comment]
Modifier keys could be fine for most users, but I don't see how the problem isn't solved by changing the tooltip position. I went ahead and tested it with [Monkey Patch](https://marketplace.visualstudio.com/items?itemName=iocave.monkey-patch) and it works just fine. I don't enable Monkey Patch in normal usage so it isn't a solution for me personally, but it does prove that changing the tooltip position solves the problem of clicking the tooltip when you're trying to click text.

And of course, requiring a modifier key to interact with tooltips/popups might solve this rather marginal problem for us, but at the cost of making the entire program less accessible for everyone. Locking it behind an opt-in pref could work, but why bother if modifying the tooltip position would mitigate the issue equally well with less cost?

Maybe you're thinking of some other use case that hasn't occurred to me. My own concern is with a scenario where you're trying to click a region of the text area that's covered up by the tooltip. If the tooltip is moved far enough away from the text that generated the tooltip, then there's no way for the tooltip to block the mouse.

Think about it like this — if the tooltip is offset from the text that generated it by 20x20px, then any attempt to click the text area underneath the tooltip would actually succeed, because it would require you to move your cursor *off of* the generating text, thereby closing the tooltip. By the time your cursor reaches the region that was covered up by the tooltip, the tooltip has already disappeared, since it was generated by text that's 20x20px away from itself.

Incredibly simple fix, it could even be implemented with CSS alone although it's probably better to go lower level so it can't bleed off the window by the offset distance amount. Modifying whatever component vs code uses to anchor the tooltips to the generating text I guess.

The exact offset might depend on how quickly the tooltip fades after the "mouseleave" event triggers on the generating text. A shorter fade duration/delay means the offset can be smaller. By that I mean the delay before the tooltip fades after the cursor leaves the generating text, not the delay before the tooltip opens.

A fade duration + delay of 0 would mean even a few pixels should work in most cases, but at that point it depends on the timing between a mouse movement in the OS and the end of a single run of the event loop, which I suppose is gonna depend on how busy the javascript engine is. So there's definitely some minimum offset required, but it can be pretty small if the tooltip fade duration + delay are shortened as much as possible.

**Comment by proimage** [comment]
@aminomancer What you're not taking into consideration is that some tooltips have elements inside that are intended to be clickable.

Let's define the issue clearly:
- Tooltips can contain links, and thus need to be clickable.
- However, tooltips also cover up/obscure code that the author may be trying to interact with.

Therefore, the solution becomes clear:
- By default, tooltips need to be mouse-phobic: they should rapidly fade to transparent when the mouse moves over them. Neither should they respond to clicks (assuming someone manages to click on them while they're fading out); instead, any clicks should be passed down to the code beneath them.
- When a modifier key is held, it removed the mouse-phobia from the tooltips, allowing them to remain on-screen when hovered, and to register mouse clicks.

The one caveat with that is that the modifier key should not be one that affects text selection; eg. if ALT-clicking makes multi-cursor selections, ALT would be a terrible modifier key to hold to make tooltips clickable. Presumably allowing a user-defined modifier key would solve this potential issue.

**Comment by trevyn** [comment]
Here’s a solution that works for me with no modifier keys:

- Tooltips have a dedicated area of the screen away from the text selection point.
- Normally, tooltip area shows tip for what is under the mouse cursor, with no delay, and no clicking.
- When the mouse cursor is over the tooltip itself (or after a keypress with no mouse movement) the tooltip area shows the tip for what is at the *text* selection point.

This keeps the tooltip area away from obscuring the text, makes links (and the color picker!) clickable, and allows very fast “scanning” of tips by just moving the mouse around.

**Comment by justsml** [comment]
Well said @proimage - Another issue that @aminomancer touched on is the issue of a delay. That alone means an offset wouldn't work in every case.

For example, if you want to click 2 or 3 lines above the cursor, just as the tooltip opens, you'll still run into the problem.

@trevyn's suggestion of a dedicated area is another good option.
I suspect it would be a huge benefit for accessibility (e.g. folks sensitive to motion, or those using non-mouse/alternate input methods, etc.)

**Comment by Sofianio** [comment]
Why is this issue still open when there is `editor.hover.sticky` setting?
And I see no mention of it in this thread.
Set it to false and the tooltip hides when hovered :)

**Comment by aminomancer** [comment]
> 
> 
> Why is this issue still open when there is `editor.hover.sticky` setting?
> And I see no mention of it in this thread.
> Set it to false and the tooltip hides when hovered :)

Weren't we just talking about how tooltips can contain hyperlinks, i.e. to MDN documentation?

**Comment by Sofianio** [comment]
@aminomancer I guess I missed that part, I never needed to click on a tooltip so it never came to my mind

**Comment by holywar20** [comment]
This is a super annoying issue. I like tool tips. I use them all the time. but VS code is especially aggressive in displaying them.

I specifically have a huge issue with them popping up while I'm typing, covering up the line above it, which I sometimes want to actually read while I'm typing. Just like an offset reposition setting that a user could configure would be huge. Either that, or like a half second delay.

Also noting issue originally filed in 2016, so apparently I'm one of the few who finds this annoying and distracting.

**Comment by proimage** [comment]
Allow me to reiterate my solution, which in all honesty I believe to be ideal:

- Tooltips are "transparent" to mouse clicks by default.
  - When a mouse click "goes through" such a tooltip, make the tooltip disappear.
- If you hold down a (user-selectable) modifier key, tooltips can be interacted with.
  - CTRL/CMD might be a good default, to allow easy copy-pasta & link clicking without having to hold down an additional modifier key.

**Comment by Jimbo99** [comment]
> I specifically have a huge issue with them popping up while I'm typing, covering up the line above it, which I sometimes want to actually read while I'm typing. Just like an offset reposition setting that a user could configure would be huge.

I would love to have the tool tips appear a little further away from the cursor, so an offset reposition would be great.

FWIW, there is already an option to delay appearance of tooltips. I use it but depending on circumstances sometimes I would like it faster and sometimes I don't want it at all. So the present delay option is far from perfect.

I think one of the problems I have is sometimes the tooltips are huge - a great chunk of a constant array or whatever that I don't want blocking the stuff I want to see. That amplifies the problem greatly.

**Comment by justsml** [comment]
@Sofianio 
> Why is this issue still open when there is `editor.hover.sticky` setting?
> And I see no mention of it in this thread.
> Set it to false and the tooltip hides when hovered :)

@Jimbo99 an offset can't ever know what lines you wish to be looking at.

#### The discussion around 'Sticky' or 'Offset' options will NEVER accommodate the multiple use-cases here.
I hope no-one is burning time building support for offset or fancier delay settings.

I've been noticing this issue more frequently as I coach & mentor people. Pair programming can be stressful anyway, but this behavior makes it border on painful. As navigator I need to be able to scan the code without asking "can you move your mouse? I was reading that..." 20x a session.
It violates a critical best practice "not to micro-manage the driver mid-flow."
It's now crystal clear to me how it harms momentum & focus for both driver & navigator.

### I did some back-of-the-napkin math from 7 hours of pairing this past week.

> The frequency of unwelcome tooltips is between 20 seconds and 4 minutes. (Defined as: anytime the users has to click around or move the mouse to get under a tooltip.)
> The majority of **delays fall between 2-15 seconds.** Sometimes minutes as the tooltip can drag focus into other parts of the app.

These small delays might be fine for some folks, however I feel my ADHD get turbocharged. I **struggle** to build momentum & get into flow. 

Here's a recording of me trying to click on methods the tooltip covers:

![VS-Code-Tooltip-bug gif-256-640](https://user-images.githubusercontent.com/397632/120232530-f6c06000-c210-11eb-9eff-9c6afdf5019b.gif)


Essentially, I don't want tooltips to be interactive or perhaps show at all *until I hold down a modifier,* like [SHIFT].

It could also be a "sticky" state you toggle by `Double Tapping [SHIFT]` or whatever the user configures.

Considering most technical folks are already familiar with `Ctrl+Click` or `Cmd+Click` to open things in a new tab. This could extend behavior in a roughly similar spirit: holding the key toggles either showing, clicking, or scrolling.

Some toggle behavior/states could be:

* show-while-modifier-key-pressed **OR**
* default to show (similar to current defaults,) but grayscale until interactivity is enabled. Clicks should passed through to underlying content as if the box isn't there.

**Comment by jordan-bucholtz** [comment]
I just want to be able to pick where the tooltip appears 😞. Lately it's been so obstructive that I'm considering going back to my vim setup, which doesn't have this issue because I'm able to select the position of the tooltip.

**Comment by Flupp** [comment]
As an inspiration I’d like to show how KDevelop deals with this problem. It combines some ideas that were already mentioned here or in other issues. I do not want to say that KDevelop has the perfect solution. However, it has an existing solution so it can be tried out to get a feeling which parts of it make sense.

Tooltips in KDevelp are accessible via hovering with the mouse and via keyboard (by pressing Alt). When using the mouse, the hover delay is also applied when switching between tooltips. Therefore the tooltip stays in place when the mouse accidentally hovers another hover target for only a short amount of time. This allows placing the tooltips farther away from the hover target to keep more context in view while still allowing mouse interactions with the tooltip.

![tooltip](https://user-images.githubusercontent.com/364471/125199864-7c9be600-e268-11eb-8925-b6e645f08995.gif)

Alternatively the tooltip content can be show in a panel. As soon as the panel is visible, all tooltip content is shown there and tooltips are disabled. The panel content follows the mouse cursor as well as the text cursor. The idea of using a panel is discussed in #51253 and was also mentioned in [#63144](https://github.com/microsoft/vscode/issues/63144#issuecomment-439273105). The [Docs View extension](https://marketplace.visualstudio.com/items?itemName=bierner.docs-view) implements this idea partially, however, this extension considers only the text cursor and only documentation therefore it cannot entirely replace the tooltips.

![toolview](https://user-images.githubusercontent.com/364471/125199913-b4a32900-e268-11eb-9705-acff577c5153.gif)

Personally, I quite like the idea of dynamically switching between showing the information in a panel or in a tooltip simply based on the visibility of the panel. Also, I think this solution might be applicable to both hover tooltips and parameter hints.

**Comment by sbmelvin** [comment]
This issue has been open for five years. I didn't know VS Code was an Atlassian product.

**Comment by isimmons** [comment]
I love having all the information popup because I'm learning right now but back off from the cursor a little bit. As @Flupp shows in Kdevelope the popups are a couple of line widths off the cursor.  vscode is literally on top of anything I'm trying to type so I can't even see what I'm typing. 

Often times the suggestion popup stays after I type a semicolon at the end of the line and then when I hit enter to go to the next line it replaces everything I just typed with whatever it was suggesting.

It seems there are several trigger keys like when I try to type it('should... for a jest test as soon as I type '('' it inserts some code that does I don't know what but starts with the letters 'it' . When I see the suggestion pop up I have to type a space to make it go away and then backspace to finish typing what I was trying to type.

 I just found the setting today to make it stop adding unwanted import statements to the top of the file that cause crashes and confusing errors.  In some cases I went back to previous code in review and just happened to see these extra import statements and couldn't figure out how they got there. Luckily in those cases the statements weren't causing any errors but in many cases they did until I figured out what was going on.

**Comment by aminomancer** [comment]
How am I STILL getting emails from this? I posted an issue a couple months ago about validation and it was fixed within less than a week without anyone else voting. Why is this particular issue taking so long to move forward?

**Comment by pflannery** [comment]
My original issue is solved. There is lot of other issues posted here that are in and around the original issue.
Going to close this off and people can open their own issues if there still not solved

**Comment by proimage** [comment]
> How am I STILL getting emails from this? I posted an issue a couple months ago about validation and it was fixed within less than a week without anyone else voting. Why is this particular issue taking so long to move forward?

To put it into perspective, when this issue was first posted, Americans were still in shock that Trump had won the 2016 elections.

**Comment by ShanConn** [comment]
These tooltips are just extremely annoying. The only things the vscode team cares about is if the product gets more addictive, and if they can use it to push users even further into azure.

**Comment by ohuu** [comment]
This issue is not fixed, please re-open it. The tooltips are incredibly annoying whether they appear immediately above or below the line your cursor is on, it would be so much better if you could set the number of lines (above or below) the cursor where any tooltip should be rendered.

Disabling tooltips is not a solution because they actually contain very useful information.

**Comment by aminomancer** [comment]
I agree the tooltips are annoying but I'm pretty sure theorizing about sinister conspiracies by microsoft isn't gonna help get this issue fixed

**Comment by JacobFV** [comment]
I tried finding the 'mouse hover wait before showing tooltip' variable in the source code but did not suceed. For anyone else who is trying to edit the code, I don't think you will have sucess changing the [`MOUSE_MOVE_MINIMUM_TIME`](https://github.com/microsoft/vscode/blob/4712e4fb0a86b2bb21c28f411191af06a8c80a45/src/vs/editor/browser/controller/mouseHandler.ts#L66) variable (in [src/vs/editor/browser/controller/mouseHandler.ts#L66](https://github.com/microsoft/vscode/blob/main/src/vs/editor/browser/controller/mouseHandler.ts#L66)).
